### ISSUE TITLE
Structure of Attention pattern mask

### ISSUE THREAD

**Issue by erlebach** [neelnanda-io/TransformerLens#284]
Hi Neel, 

In the clean transformer demo, the mask pattern for attention is generated with the command 
```
        mask = torch.triu(torch.ones(attn_scores.size(-2), attn_scores.size(-1), device=attn_scores.device), diagonal=1).bool()
```
which defines the mask as all elements of the upper triangular matrix, *excluding* the first upper diagonal. My question is the following: why are you excluding the first upper diagonal? I checked that other implementations on the web do not do this. I recall in your walkthrough that you were surprised that your model simply duplicated your test set. Could this have been the reason? Thanks.  

   Gordon

**Comment by jbloomAus** [comment]
@erlebach  Is this just tutorial code? If so the tutorial code is likely broken and it's a bug. Can you provide screen shots of the attention map comparison? Should be pretty clearly wrong if what you say is true.

**Comment by neelnanda-io** [comment]
No, this is correct. The mask is the tokens the model is *not *allowed to
attend between. The first diagonal is just the main diagonal, and the model
is allowed to attend from a token to itself. But this is easy to test, just
look at the model's attention patterns on a few inputs and check if they
have illegal attention

**Comment by jbloomAus** [comment]
Right. I interpreted first upper as the next token. Not an issue then. I'll
close this.

**Comment by erlebach** [comment]
Thanks, I figured it out. :-)
